src/include/denc.h

```diff
@@ -307,6 +349,14 @@ inline void denc_signed_varint(T& v, bufferptr::iterator& p)
 inline void denc_varint_lowz(uint64_t v, size_t& p) {
   p += sizeof(v) + 2;
 }
+
+inline void denc_varint_lowz_exact(uint64_t v, size_t& p) {
+  int lowznib = v ? (ctz(v) / 4) : 0;
+  if (lowznib > 3)
+    lowznib = 3;
+  v >>= lowznib * 4 - 2;
```

Comment:
Superscalar, out-of-order microarchitectures are with us for decades. Piotr, I can sure you that nobody here even tries to think about in-order ones as nobody is going to deploy BlueStore + super-fast SSDs on Atoms. That was the exact reason why instructions' **throughputs** (not latencies!) have been quoted previously. I believe those data are still valid. Please take a look on the note on page 647 of the Intel's Optimization Reference Manual:

> Although instruction latency may be useful in some limited situations (e.g., a tight loop with a dependency chain that exposes instruction latency), **software optimization on super-scalar, out-of-order microarchitecture, in general, will benefit much more on increasing the effective throughput** of the larger-scale code path.

I feel we still have a misunderstanding here, so let me bring more context to the discussion. `denc_varint_lowz_exact` and familiar would rarely operate alone. Instead, they will be combined, possibly multiple times, into blocks to process structures. Such constructs can be inlined and hopefully result in islands of well-parallelizable code. In other words: we can obtain multiple dependency chains that can be executed together thanks to the huge out-of-order window (around hundred of instructions per thread) of nowadays CPUs. However, to achieve high IPC ratio the whole pipeline should be constantly saturated. Non-optimal instruction selection can hurt not only at the micro-op execution stage but instruction decode as well:

> There are four decoders, which can handle instructions generating one or more µops according to certain patterns. (...) **Instructions that generate 3 or 4 µops are decoded alone.** [1]

Of course, there are measures to minimise such penalties. As you mentioned, compilers can carefully reorder instructions. Modern x86 incarnations are able to bypass decoders and pull micro-ops directly from a dedicated cache. They can also smooth front-end bubbles in the micro-op queue or reorder micro-ops later at ROB/Reservation Station, and so on. Still, the best strategy to avoid costs of wrongly chosen instructions is to simply pick up the proper ones if possible.

I have a feeling that I haven't stressed enough that the intended optimization isn't narrowed to the single line. As this might be the source of the misunderstanding, let me describe algorithm I had in mind while mentioning `denc_lba_exact` in the initial comment:

1. Handle the special case of `v == 0`.
2. Calculate `nbits = cbits(v)` and `ntzeros = ctz(v)`. IIRC these operations can be taken effectively even on i386 basing on `BSR` and `BSF` instructions.
3. Saturate the `ntzeros` to `12`. Compilers are supposed to implement that with `CMOV` instead of a conditional branch.
4. Zeroize two LSBs of the `ntzeros`.
5. Finally execute `(nbits - ntzeros + 2 + 6) / (CHAR_BIT - 1)`. As the divider is known at compile time, any reasonable compiler should use multiplication and non-variable right shift instead.
 
On machines offering the BMI1 extension we could use `LZCNT` and `TZCNT` at the point 2. As these instructions provide well-defined results for the `v == 0` case, we would get an opportunity to remove the very last conditional branch and set `nbits` to e.g. `12` with `CMOV` instead. However, I doubt whether branching is an issue here. Even in the current, non-optimal implementation of  `denc_varint_lowz_exact` I cannot spot a problem's pattern resembling the one you've fixed in the referenced PR.

Before moving forward, I would like to congratulate the outstanding work you made at #10935! Though, even there I would be careful with blaming the conditional branching per se. Sure, `crush_ln()` had a tight loop possibly exposing the latencies. On the other hand, the function was pretty short. Estimating very, very roughly it had around 100 instructions in the worst case. In such scenario the loop itself was responsible for ~70 of them. You've reasonably replaced it with just a couple of instructions achieving ~3x speed-up. I think we would need to take a look on PMU's counters to be conclusive.

Finally, I must disagree that conditional branches are *always* expensive. I would rather say *it depends*. Many factors should be taken into account: whether a branch was mispredicted or not, surrounded by many non-branch instructions or not (the tight loop case), direct or indirect, near (without crossing the cache line/page boundaries) or far to just name few of them. Total costs might vary from negligible (predicted, local and interleaved with many other instructions -- narrowed basically to the instructions' cost) to devastating (far branches causing cache and ITLB/STLB misses and thus page walks, page faults, ...).
It's also worth to add that CPUs these days are really good in making these predictions. I can recall that @aclamk took a closer look on that during his work on [the intrusive, per-scope PMU statistics gatherer](https://github.com/aclamk/pmu). Though dozens years in the business, he was impressed on how those silicon x86 beasts are good in cheating the barriers!

Anyway, the `*_exact` family of functions would be used in BlueStore only. At the moment we have much more severe headaches with providing CPU with data. When it comes to focus on `denc.h` we would be very happy people.

Regards,
Radek

[1]. Agner Fog, ["The microarchitecture of Intel and AMD CPUs"](http://www.agner.org/optimize/microarchitecture.pdf), page 121